cpu/nrf52/radio/nrf802154/nrf802154.c

```diff
@@ -0,0 +1,381 @@
+/*
+ * Copyright (C) 2017 Freie Universität Berlin
+ *               2017 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_nrf52
+ * @{
+ *
+ * @file
+ * @brief       Implementation of the radio driver for nRF52 radios
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Dimitri Nahm <dimitri.nahm@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include <string.h>
+#include <errno.h>
+
+#include "cpu.h"
+#include "luid.h"
+#include "mutex.h"
+
+#include "net/ieee802154.h"
+#include "net/netdev/ieee802154.h"
+#include "nrf802154.h"
+
+#define ENABLE_DEBUG    (1)
+#include "debug.h"
+
+
+/**
+ * @brief   Possible internal device states
+ */
+typedef enum {
+    STATE_OFF,                  /**< device is powered off */
+    STATE_IDLE,                 /**< device is in idle mode */
+    STATE_RX,                   /**< device is in receive mode */
+} idle_state_t;
+
+
+static const netdev_driver_t nrf802154_netdev_driver;
+
+netdev_ieee802154_t nrf802154_dev = {
+    {
+        .driver = &nrf802154_netdev_driver,
+        .event_callback = NULL,
+        .context = NULL,
+    #ifdef MODULE_NETSTATS_L2
+        .stats = { 0 }
+    #endif
+    },
+#ifdef MODULE_GNRC
+#ifdef MODULE_GNRC_SIXLOWPAN
+    .proto = GNRC_NETTYPE_SIXLOWPAN,
+#else
+    .proto = GNRC_NETTYPE_UNDEF,
+#endif
+#endif
+    .pan = IEEE802154_DEFAULT_PANID,
+    .short_addr = { 0, 0 },
+    .long_addr = { 0, 0, 0, 0, 0, 0, 0, 0 },
+    .chan = IEEE802154_DEFAULT_CHANNEL,
+    .flags = 0
+};
+
+static uint8_t rxbuf[IEEE802154_FRAME_LEN_MAX + 2]; /* length PHR + PSDU + LQI */
+static uint8_t txbuf[IEEE802154_FRAME_LEN_MAX + 2]; /* length PHR + PSDU + LQI */
+
+static mutex_t txlock;
+
+static volatile idle_state_t idle_state = STATE_RX;
+
+/**
+ * @brief   Set radio into idle (DISABLED) state
+ */
+static void go_idle(void)
+{
+    /* set device into basic disabled state */
+    NRF_RADIO->EVENTS_DISABLED = 0;
+    NRF_RADIO->TASKS_DISABLE = 1;
+    while (NRF_RADIO->EVENTS_DISABLED == 0) {}
```

Comment:
Maybe for debugging purposes add a counter and an assertion in order to prevent infinite loops.